apis/python/src/tiledbsoma/collection.py

```diff
@@ -102,81 +117,41 @@ def _not_implemented(self, *args: Any, **kwargs: Any) -> NoReturn:
-        )
-
-    @property
-    def _tiledb_obj(self) -> tiledb.Group:
-        "get the open tiledb.Group handle"
-        assert self._open_tiledb_group is not None  # => not self.closed
-        return self._open_tiledb_group
-
-    def close(self) -> None:
-        self._open_tiledb_group = None
-        super().close()  # closes self._open_tiledb_group via self._close_stack
-
     def __len__(self) -> int:
         """
         Return the number of members in the collection
         """
-        self._load_tdb_group_cache()
-        return 0 if self._cached_values is None else len(self._cached_values)
+        return len(self._group_contents)
 
     def __getitem__(self, key: str) -> CollectionElementType:
         """
         Gets the value associated with the key.
         """
-        err_str = f"{self.__class__.__name__} has no attribute '{key}'"
 
-        # Load cached TileDB Group if not yet loaded or if key not in cache.
-        if self._cached_values is None or key not in self._cached_values:
-            self._load_tdb_group_cache()
-            if self._cached_values is None:
-                # This collection was not yet created
-                # TODO: SOMA needs better exception types
-                raise DoesNotExistError("Collection has not been created")
+        err_str = f"{self.__class__.__name__} has no item {key!r}"
 
-        # if element is in the TileDB Group, so make a SOMA in-memory object to represent it.
-        if key in self._cached_values:
-            soma = self._cached_values[key].soma
-            if soma is None:
-                from .factory import _construct_member
-
-                tdb = self._cached_values[key].tdb
-                soma = _construct_member(tdb.uri, self.context, tdb.type)
-                if soma is None:
-                    # if we were unable to create an object, it wasn't actually a SOMA object
-                    raise KeyError(err_str)
-
-                self._cached_values[key].soma = soma
-
-            return cast(CollectionElementType, self._cached_values[key].soma)
-
-        raise KeyError(err_str)
+        try:
+            entry = self._group_contents[key]
+        except KeyError:
+            raise KeyError(err_str) from None
+        if entry.soma is None:
+            from . import factory  # Delayed binding to resolve circular import.
+
+            if entry.cls is tiledb.Array:
+                opener: Callable[..., TDBHandle] = factory._open_array
+            elif entry.cls is tiledb.Group:
+                opener = factory._open_group
+            else:
+                raise SOMAError(
+                    f"internal error: unrecognized group entry type {entry.cls}"
+                )
+            entry.soma = opener(
+                uri=entry.uri,
+                mode=self.mode,
+                context=self.context,
+                soma_type=None,
+            )
```

Comment:
What's later going to close the thing? (If it happens implicitly and I missed it, then it's probably worth an explanatory comment?)